In [2]:
# file from this page https://www.swisstopo.admin.ch/de/amtliches-ortschaftenverzeichnis#Download
# url: https://data.geo.admin.ch/ch.swisstopo-vd.ortschaftenverzeichnis_plz/ortschaftenverzeichnis_plz/ortschaftenverzeichnis_plz_4326.csv.zip
# 

In [3]:
import pandas as pd
import numpy as np
import zipfile

In [4]:
zip_path = "data/ortschaftenverzeichnis_plz_4326.csv.zip" 

with zipfile.ZipFile(zip_path) as z:
    print(z.namelist())


['AMTOVZ_CSV_WGS84/', 'AMTOVZ_CSV_WGS84/AMTOVZ_CSV_WGS84.csv']


In [5]:
with zipfile.ZipFile(zip_path) as z:
    with z.open("AMTOVZ_CSV_WGS84/AMTOVZ_CSV_WGS84.csv") as f:
        plz_df = pd.read_csv(f, sep=";", encoding="utf-8-sig")

plz_df.head()


,Ortschaftsname,PLZ4,Zusatzziffer,ZIP_ID,Gemeindename,BFS-Nr,Kantonskürzel,Adressenanteil,E,N,Sprache,Validity
0,Lausanne 25,1000,25,125,Lausanne,5586,VD,100 %,6.682261,46.559589,fr,2008-07-01
1,Lausanne 26,1000,26,126,Lausanne,5586,VD,100 %,6.695932,46.554570,fr,2008-07-01
2,Lausanne 27,1000,27,127,Lausanne,5586,VD,100 %,6.681672,46.542070,fr,2008-07-01
3,Lausanne,1003,0,150,Lausanne,5586,VD,100 %,6.631877,46.520577,fr,2008-07-01
4,Lausanne,1004,0,151,Lausanne,5586,VD,100 %,6.618649,46.528291,fr,2008-07-01


In [6]:
plz_df.tail()

,Ortschaftsname,PLZ4,Zusatzziffer,ZIP_ID,Gemeindename,BFS-Nr,Kantonskürzel,Adressenanteil,E,N,Sprache,Validity
5713,Alt St. Johann,9656,0,5482,Wildhaus-Alt St. Johann,3359,SG,100 %,9.273176,47.190242,de,2008-07-01
5714,Unterwasser,9657,0,5483,Wildhaus-Alt St. Johann,3359,SG,100 %,9.310688,47.198959,de,2008-07-01
5715,Wildhaus,9658,0,5484,Wildhaus-Alt St. Johann,3359,SG,99.432 %,9.344029,47.201786,de,2008-07-01
5716,Wildhaus,9658,0,5484,Grabs,3273,SG,0.474 %,9.338930,47.173097,de,2008-07-01
5717,Wildhaus,9658,0,5484,Gams,3272,SG,0.095 %,9.391822,47.201209,de,2008-07-01


In [7]:
plz_df = plz_df.rename(columns={
    "Ortschaftsname": "locality_name",
    "PLZ4": "postal_code",
    "Zusatzziffer": "postcode_suffix",
    "Gemeindename": "municipality_name",
    "BFS-Nr": "municipality_id",
    "Kantonskürzel": "canton_code",
    "Adressenanteil": "address_share_pct",
    "Sprache": "language",
    "E": "longitude",
    "N": "latitude",
})

plz_df.head(20)


,locality_name,postal_code,postcode_suffix,ZIP_ID,municipality_name,municipality_id,canton_code,address_share_pct,longitude,latitude,language,Validity
0,Lausanne 25,1000,25,125,Lausanne,5586,VD,100 %,6.682261,46.559589,fr,2008-07-01
1,Lausanne 26,1000,26,126,Lausanne,5586,VD,100 %,6.695932,46.554570,fr,2008-07-01
2,Lausanne 27,1000,27,127,Lausanne,5586,VD,100 %,6.681672,46.542070,fr,2008-07-01
3,Lausanne,1003,0,150,Lausanne,5586,VD,100 %,6.631877,46.520577,fr,2008-07-01
4,Lausanne,1004,0,151,Lausanne,5586,VD,100 %,6.618649,46.528291,fr,2008-07-01
5,Lausanne,1005,0,152,Lausanne,5586,VD,100 %,6.642754,46.520897,fr,2008-07-01
6,Lausanne,1006,0,153,Lausanne,5586,VD,100 %,6.633803,46.512854,fr,2008-07-01
7,Lausanne,1007,0,154,Lausanne,5586,VD,100 %,6.613568,46.517918,fr,2008-07-01
8,Prilly,1008,0,156,Prilly,5589,VD,100 %,6.604546,46.536753,fr,2008-07-01
9,Jouxtens-Mézery,1008,2,5852,Jouxtens-Mézery,5585,VD,100 %,6.595169,46.549512,fr,2008-07-01


In [8]:
plz_df.isna().sum()

locality_name         0
postal_code           0
postcode_suffix       0
ZIP_ID                0
municipality_name     0
municipality_id       0
canton_code          20
address_share_pct     0
longitude             0
latitude              0
language              0
Validity              0
dtype: int64

In [9]:
plz_df[plz_df.isna().any(axis=1)]

,locality_name,postal_code,postcode_suffix,ZIP_ID,municipality_name,municipality_id,canton_code,address_share_pct,longitude,latitude,language,Validity
5584,Nendeln,9485,0,5389,Eschen,7007,NaN,99.054 %,9.543433,47.197430,de,2008-07-01
5585,Nendeln,9485,0,5389,Gamprin,7009,NaN,0.946 %,9.547856,47.195783,de,2008-07-01
5586,Schaanwald,9486,0,5390,Mauren,7008,NaN,100 %,9.561912,47.211973,de,2008-07-01
5587,Gamprin-Bendern,9487,0,5391,Gamprin,7009,NaN,99.507 %,9.511399,47.220056,de,2008-07-01
5588,Gamprin-Bendern,9487,0,5391,Eschen,7007,NaN,0.329 %,9.503241,47.196817,de,2008-07-01
5589,Gamprin-Bendern,9487,0,5391,Vaduz,7001,NaN,0.164 %,9.508389,47.193144,de,2008-07-01
5590,Schellenberg,9488,0,5392,Schellenberg,7011,NaN,100 %,9.544028,47.230306,de,2008-07-01
5591,Vaduz,9490,0,5393,Vaduz,7001,NaN,100 %,9.517533,47.143139,de,2008-07-01
5592,Ruggell,9491,0,5394,Ruggell,7010,NaN,100 %,9.525920,47.238815,de,2008-07-01
5593,Eschen,9492,0,5395,Eschen,7007,NaN,100 %,9.521684,47.212177,de,2008-07-01


In [10]:
plz_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5718 entries, 0 to 5717
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   locality_name      5718 non-null   str    
 1   postal_code        5718 non-null   int64  
 2   postcode_suffix    5718 non-null   int64  
 3   ZIP_ID             5718 non-null   int64  
 4   municipality_name  5718 non-null   str    
 5   municipality_id    5718 non-null   int64  
 6   canton_code        5698 non-null   str    
 7   address_share_pct  5718 non-null   str    
 8   longitude          5718 non-null   float64
 9   latitude           5718 non-null   float64
 10  language           5718 non-null   str    
 11  Validity           5718 non-null   str    
dtypes: float64(2), int64(4), str(6)
memory usage: 536.2 KB


In [11]:
# convert column address_share_pct to number
plz_df["address_share_pct"] = (
    plz_df["address_share_pct"].str.replace("%", "").str.strip().astype(float)
)

plz_df["address_share_pct"].head()


0    100.0
1    100.0
2    100.0
3    100.0
4    100.0
Name: address_share_pct, dtype: float64

In [12]:
weight_df = (
    plz_df.groupby(["postal_code"])["address_share_pct"].sum()
    / plz_df.groupby("postal_code")["ZIP_ID"].nunique()
).dropna().rename("weight_pct").reset_index()

weight_df.head()


,postal_code,weight_pct
0,1000,100.0
1,1003,100.0
2,1004,100.0
3,1005,100.0
4,1006,100.0


In [13]:
weight_df[weight_df['weight_pct'] != 100]

,postal_code,weight_pct
137,1212,100.0010
141,1216,100.0010
152,1228,100.0010
298,1453,100.0005
381,1628,100.0010
...,...,...
3108,9469,100.0010
3150,9536,99.9990
3156,9548,100.0010
3174,9615,100.0010


In [14]:
#check meaning of address_share_pct
plz_pct = plz_df.groupby(["postal_code"])["address_share_pct"].sum().reset_index()
plz_pct

,postal_code,address_share_pct
0,1000,300.000
1,1003,100.000
2,1004,100.000
3,1005,100.000
4,1006,100.000
...,...,...
3185,9652,100.000
3186,9655,100.000
3187,9656,100.000
3188,9657,100.000


In [15]:
count_zip = plz_df.groupby("postal_code")["ZIP_ID"].nunique().reset_index()
count_zip

,postal_code,ZIP_ID
0,1000,3
1,1003,1
2,1004,1
3,1005,1
4,1006,1
...,...,...
3185,9652,1
3186,9655,1
3187,9656,1
3188,9657,1


In [16]:
merge_df = pd.merge(
    plz_pct,
    count_zip,
    on='postal_code',
    how="left"
)

merge_df

,postal_code,address_share_pct,ZIP_ID
0,1000,300.000,3
1,1003,100.000,1
2,1004,100.000,1
3,1005,100.000,1
4,1006,100.000,1
...,...,...,...
3185,9652,100.000,1
3186,9655,100.000,1
3187,9656,100.000,1
3188,9657,100.000,1


In [17]:
# check null on merge
merge_df.isna().sum()

postal_code          0
address_share_pct    0
ZIP_ID               0
dtype: int64

In [18]:
# check sum = 100 in order to validate pct when analysis
merge_df['check_pct'] = merge_df['address_share_pct']/merge_df['ZIP_ID']

merge_df.head()

,postal_code,address_share_pct,ZIP_ID,check_pct
0,1000,300.0,3,100.0
1,1003,100.0,1,100.0
2,1004,100.0,1,100.0
3,1005,100.0,1,100.0
4,1006,100.0,1,100.0


In [19]:
# check postal_code != 100
merge_df[merge_df['check_pct'] != 100].sort_values('check_pct', ascending=False)

,postal_code,address_share_pct,ZIP_ID,check_pct
137,1212,100.001,1,100.001
152,1228,100.001,1,100.001
381,1628,100.001,1,100.001
708,2505,100.001,1,100.001
887,3068,100.001,1,100.001
...,...,...,...,...
3022,9107,99.999,1,99.999
3062,9308,99.999,1,99.999
3084,9425,99.999,1,99.999
3077,9405,99.999,1,99.999


In [20]:
merge_df[merge_df['check_pct'] != 100].sort_values('check_pct', ascending=True)

,postal_code,address_share_pct,ZIP_ID,check_pct
439,1700,99.999,1,99.999
885,3066,99.999,1,99.999
1081,3513,99.999,1,99.999
1780,6016,99.999,1,99.999
1855,6170,99.999,1,99.999
...,...,...,...,...
3050,9242,100.001,1,100.001
3091,9436,100.001,1,100.001
3108,9469,100.001,1,100.001
3174,9615,100.001,1,100.001


In [21]:
# check each municipality (key for mappping with incometax data)
weight_df = (
    plz_df.groupby(["postal_code", "municipality_id"])["address_share_pct"].sum()
    / plz_df.groupby("postal_code")["ZIP_ID"].nunique()
).dropna().rename("weight_pct").reset_index()

weight_df.head(10)


,postal_code,municipality_id,weight_pct
0,1000,5586,100.0
1,1003,5586,100.0
2,1004,5586,100.0
3,1005,5586,100.0
4,1006,5586,100.0
5,1007,5586,100.0
6,1008,5585,50.0
7,1008,5589,50.0
8,1009,5590,100.0
9,1010,5586,100.0


In [22]:
check = weight_df.groupby("postal_code")["weight_pct"].sum()
print((~check.between(99.9, 100.1)).sum(), "out of", len(check))


0 out of 3190


In [23]:
muni_lookup = plz_df[["municipality_id", "municipality_name", "canton_code"]].drop_duplicates(subset="municipality_id")

weight_df = weight_df.merge(muni_lookup, on="municipality_id", how="left")
weight_df.head(10)


,postal_code,municipality_id,weight_pct,municipality_name,canton_code
0,1000,5586,100.0,Lausanne,VD
1,1003,5586,100.0,Lausanne,VD
2,1004,5586,100.0,Lausanne,VD
3,1005,5586,100.0,Lausanne,VD
4,1006,5586,100.0,Lausanne,VD
5,1007,5586,100.0,Lausanne,VD
6,1008,5585,50.0,Jouxtens-Mézery,VD
7,1008,5589,50.0,Prilly,VD
8,1009,5590,100.0,Pully,VD
9,1010,5586,100.0,Lausanne,VD


In [24]:
weight_df.isna().sum()

postal_code           0
municipality_id       0
weight_pct            0
municipality_name     0
canton_code          20
dtype: int64

In [25]:
#export

weight_df.to_csv("data/postal_code_municipality_mapping.csv", index=False)
print("saved:", len(weight_df), "rows")


saved: 4985 rows


The weighting assumes every locality (`ZIP_ID`) within a postal code has roughly equal
  number of addresses — the source file gives no actual size/count per locality, so this is
  the best approximation available without an additional data source.
- 20 rows (postal codes 9485–9498) belong to **Liechtenstein**, not Switzerland — `canton_code`
  is missing for these since Liechtenstein has no Swiss canton. They will not match any row in
  the Swiss tax-statistics files (`num_taxpayers_by_gemeinde.csv` / `total_income_by_gemeinde.csv`).

#--by Lavanya

In [26]:
# Original plz_df also contains latitude and longitude. Export one additional postcode-coordinate file.
# Create one representative coordinate for each postal code
postcode_coordinates_df = (
    plz_df
    .groupby("postal_code", as_index=False)
    .agg(
        postcode_latitude=("latitude", "mean"),
        postcode_longitude=("longitude", "mean"),
        locality_name=("locality_name", "first")
    )
)

postcode_coordinates_df["postal_code"] = (
    postcode_coordinates_df["postal_code"]
    .astype(str)
    .str.zfill(4)
)

postcode_coordinates_df.to_csv(
    "data/postcode_coordinates.csv",
    index=False
)

print("Saved postcode coordinates:", postcode_coordinates_df.shape)
postcode_coordinates_df.head()

Saved postcode coordinates: (3190, 4)


,postal_code,postcode_latitude,postcode_longitude,locality_name
0,1000,46.552076,6.686622,Lausanne 25
1,1003,46.520577,6.631877,Lausanne
2,1004,46.528291,6.618649,Lausanne
3,1005,46.520897,6.642754,Lausanne
4,1006,46.512854,6.633803,Lausanne
